### PR TITLE
Add support for file-based environment variables in environment-to-ini

### LINKED ISSUE #19856: Add support for file-based environment variables
### Feature Description

In the new container-centric kubernetes-oriented world, storing secrets (or other related data) inside the ini settings file as plain text could be a major security threat and could lead to secrets end up being pushed to a git repository. And managing secrets as plain text in environment variables could also represent a security threat in setups that are using monitoring/reporting/orchestrating tools that log and store environment variables. That's why docker/Kubernetes and other tools offer the ability to manage secrets as separate artifacts in a secure way (on-disk encrypted, plain-text inside the container).

When using `docker secret` or other tools like Kubernetes, HashiCorp Vault, ... and were the secret is mounted as a file in `/run/secrets/<SECRET_NAME>` there is a need for a mechanism to load those secrets from the mounted-plain-text-files to make them available to gitea.

The proposal is that any settings in `app.ini` can be set or overridden with the content of a file by defining an environment variable of the form: `GITEA__section_name__KEY_NAME__FILE` that points to a file (in case of `docker secrets` it could point to `/run/secrets/<SECRET_NAME>`).

This technique could be used in `docker-compose.yml` as follows:
```dockerfile
version: '3.9'
services:
  server:
    container_name: gitea
    image: gitea/gitea:1.16.8
    environment:
      USER_UID: 1000
      USER_GID: 1000
      # Database
      GITEA__database__DB_TYPE: postgres
      GITEA__database__HOST: gitea-db:5432
      GITEA__database__NAME: gitea
      GITEA__database__USER: gitea
      GITEA__database__PASSWD__FILE: /run/secrets/gitea_db_password
    secrets:
    - gitea_db_password
    networks:
    - gitea-net
    
  db:
    container_name: gitea-db
    image: postgres:14-alpine
    environment:
      POSTGRES_DB: gitea
      POSTGRES_USER: gitea
      POSTGRES_PASSWORD_FILE: /run/secrets/gitea_db_password
    secrets:
      - gitea_db_password
    networks:
      - gitea-net

secrets:
  gitea_db_password:
    external: true
```

As discussed with @zeripath in #19858 I will summarize the usability and security implications of the discussed implementations:
*:green_circle: for positive point, :red_circle: for negative point and empty when N/A or could be prevented*

1. Save secrets in ini file (**Current implementation**)
2. Save secrets in overlaid dedicated ini file
3. Make every key have an optional __FILE ending which would automatically read from a file.
   a. Implemented inside `gitea`: Load secrets from files to memory.
   b. Implemented inside `environment-to-ini`: Load secrets from file and populate ini file inside the container. #19857
5. Make a dedicated secrets section in init file and load from files to memory.

| Implication  | 1 (Current) | 2 | 3.a | 3.b #19857 | 4 |
| --- | --- | --- | --- | --- | --- |
| Secrets as plain text in the host  | :red_circle:  | | | | |
| Secrets encrypted in the host (docker secret, Kubernetes, HashiCorp Vault, ...)  |  | :orange_circle:   | :green_circle: | :green_circle: | :green_circle:  |
| Secrets grouped with the other settings  | :red_circle:  | | | | |
| Secrets separated from the other settings[\*]  |  | :green_circle: | :green_circle: | :green_circle: | :green_circle: |
| Secrets as plain text inside the container[\*\*]  | :red_circle:  | :red_circle: | :red_circle: | :red_circle: | :red_circle: |
| Secrets as plain text in volumes[\*\*\*]  | :yellow_circle:  | | | :yellow_circle: | |
| Secrets re-usable with other containers (wide usage)  | | | :green_circle: | :green_circle: | :green_circle: |
| Major update to ini file  | | | | | :red_circle: |
| Update only  environment-to-ini | | | | :purple_circle:  | |

:orange_circle: : Managing the whole overlaid ini file (dedicated for secrets) with `docker secret`
:purple_circle: : I don't know if it could be a positive or negative point (in the short run) but the risk of breaking gitea is reduced and, for sure, for the long run it is better to implement it gitea-side.
[\*]: When the secrets are separated from the general settings they could be managed by `docker secret` and mounted only when running the container.
[\*\*]: In all cases the secrets mounted as files will be present as plain text inside the container.
[\*\*\*]: Having secrets as plain text in volumes could be a security risk but could be mitigated if using an encrypted filesystem and setting the correct file-permission for the volume directory to prevent unprivileged processes from accessing it.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by aminosbh**
Improve `environment-to-ini` to allow for file content to be set as the value of an environment variable.
Useful when using `docker secret` and were the secret is mounted as a file in `/run/secrets/<SECRET_NAME>`.

Any settings in `app.ini` can be set or overridden with the content of a file by defining an environment variable of the form: `GITEA__section_name__KEY_NAME__FILE` that points to a file.

Fixes #19856

### CODE REVIEW

**Comment by KarolNi** [comment]
Fixes https://github.com/go-gitea/gitea/issues/10311

**Comment by nesono** [comment]
Any reason why this PR was discontinued? This change would be extremely useful imho.

**Comment by Doddophonique** [comment]
I agree, this is a key functionality for a lot of production environments. Having it embedded in the code is far better than implementing potentially unsafe workarounds.

**Comment by lunny** [comment]
Maybe another prefix like `GITEA__FILE__` is better than suffix.

**Comment by nesono** [comment]
> Maybe another prefix like `GITEA__FILE__` is better than suffix.

I found suffixed version to be quite standard with some other Docker containers, but I don't have a strong opinion. Could you land this change for us, @lunny ? It looks like you are a member.

Then, I was writing my own wrapper now, based a previous issue that was discussed here, which works as a workaround for me for now. Here's the code for completion (I modified the original prosposal to be compliant with `/bin/sh` syntax.

```bash
# This file has been copied from the original Docker container, as follows:
# docker run gitea/gitea:1.18.0
# docker exec <container_name>> cat /usr/bin/entrypoint > gitea.entrypoint.sh

# The patch starts here. It's a rewritten form of what was posted here
# https://github.com/go-gitea/gitea/issues/10311.
# >>> SNIP
export_secret_as_env_var()
{
    secret=$1
    envFile="${secret}_FILE"
    envFileName="$(printenv "${envFile}")"
    if [ -n "${envFileName}" ]; then
        if [ -f "${envFileName}" ]; then
          val=$(cat "${envFileName}")
          export "${secret}"="$val"
          echo "${secret} environment variable was set via secret ${envFileName}"
        else
          >&2 echo "Error: Secret ${secret} cannot be set via secret ${envFileName}. Not a file"
        fi
    else
        echo "Warn: ${secret} environment variable ist not defined in secret"
    fi
}

# Set environment variables by their respective secrets
export_secret_as_env_var "GITEA__database__PASSWD"
export_secret_as_env_var "GITEA__database__USER"
export_secret_as_env_var "GITEA__mailer__USER"
export_secret_as_env_var "GITEA__mailer__PASSWD"
# <<< SNAP
```

It seemed though, that the `GITEA__database__PASSWD` was already supported, maybe that's worthwhile to check.
I think a username should be protected, too, though. Adds yet another layer, even if it's small.

Cherrs and thanks for reading this far :)

**Review by dynamicat — COMMENTED** [review]
Can someone review approve and merge this?

File: contrib/environment-to-ini/environment-to-ini.go [hunk 1]
```diff
@@ -137,6 +149,21 @@ func runEnvironmentToIni(c *cli.Context) error {
...
+			if isFile {
+				if content, err := os.ReadFile(value); err == nil {
+					value = string(content)
+				} else {
+					log.Fatal("Failed to load value from file '%s': %v", value, err)
+				}
+			} else {
+				log.Fatal("File '%s' not found", value)
+			}
+		}
```
Comment (wxiaoguang) [inline contrib/environment-to-ini/environment-to-ini.go:166]:
```suggestion
		if isFileBased {
			if content, err := os.ReadFile(value); err == nil {
				value = string(content)
			} else {
				log.Fatal("Failed to load value from file %q: %v", value, err)
			}
		}
```

I guess this is enough?

**Comment by wxiaoguang** [comment]
Although there are some conflicts, I think this PR could be fine-tuned and merged.

Could maintainers with writer permission do some helps?

**Comment by silverwind** [comment]
Generally I think that what `environment-to-ini` does needs to move to the main application so that environment variables are recognized everywhere, not just with the docker container because docker is just one of many deployment option. If someone wants to use systemd with env vars, they should be able to.

**Comment by wxiaoguang** [comment]
Replaced by  Make environment-to-ini support loading key value from file #24832

**Pull request closed without merging**